Modules/Loadable/SubjectHierarchy/Widgets/qSlicerSubjectHierarchyExportPlugin.cxx

```diff
@@ -43,6 +43,31 @@
 // MRML includes
 #include <vtkMRMLStorableNode.h>
 
+//-----------------------------------------------------------------------------
+// Utility function for getting the subject-hierarchy-based relative path of an item.
+// Starting from fromItemID, this function climbs up the subject hierarchy tree until it reaches toItemID,
+// and it constructs a list of item names that it encounters along the way. The list goes into the return parameter `path`.
+// The list can be thought of as the interval "(fromItem, toItem]"; i.e. it excludes fromItem.
+// The returned bool indicates success.
+static bool getSubjectHierarchyPath(vtkMRMLSubjectHierarchyNode* shNode, vtkIdType fromItemID, vtkIdType toItemID, QStringList& path)
+{
+  path.clear();
+  vtkIdType endItemID = shNode->GetItemParent(toItemID); // One item above the highest we want to go
+  int iteration = 0;
+  for (vtkIdType itemID = shNode->GetItemParent(fromItemID); itemID != endItemID; itemID = shNode->GetItemParent(itemID))
+    {
+    if (itemID==0 || iteration > 9999)
```

Comment:
It would be much more clear to do an infinite loop detection more explicitly. Since the IDs are numbers and we expect only a couple of them (hierarchy is not expected to be deeper) they can be compared very quickly, so it would be a negligible overhead to put the IDs in an array (not a list or map, because for small structures array is expected to have lower overhead) and before each insertion check if the item is already in it.

If you are concerned about the performance then you can do what you do now with a limit of 100 (hierarchy should be really never go above that) and if that number is exceeded then start again and store the IDs and detect the loop and if you find that indeed there is a loop then report the error (with details, such as the items in the loop).

Reply:
Explicitly: check each item ID if it is in the parent list already; if it is then we know that there is an infinite loop.
Implicit: set a limit for the maximum number of parent items (e.g., total number of items + 1), if that number is reached then we know that there must have been an infinite loop.

> I'm not putting the IDs in a container,

Not in the current implementation, but it would be very easy to change this. Since we never expect to get into an infinite loop, it makes sense to build both the string list (this will be returned) and ID list (for quick loop check) at the same time.

Reply:
> Explicitly: check each item ID if it is in the parent list already; if it is then we know that there is an infinite loop.

Ah I see what you mean, but this approach misses some types of infinite loops: an infinite loop where the item ID returned is always different. Okay that's technically not possible because there are only finitely many possible vtkIDType values, but the number could be very large depending on the size of a vtkIDType, so we could still end up with an effectively stalled program.

When protecting against stalling the program, I don't want to make assumptions about the size of vtkIDType and I don't want to make assumptions about whether item ids will always be delivered correctly, i.e. from a small pool of possible ids.

> and ID list (for quick loop check) 

The ID list seems like a lot more overhead for loop check compared to just an integer counter. I'm not sure I'm seeing the advantage of keeping an ID list just for loop check. With my current method, the loop check is an integer comparison, versus iterating through an ID array and doing a bunch of equality checks.

(Sorry if I am still misunderstanding your suggestion!)

Reply:
Since most cases you will have about 2-3 values in that array, it will be plenty fast. Maybe in rare cases it will be as many as 5 values. Iterating through 5 integer values in an array is nothing compared to string manipulations and file IO.

In the extremely rare case of an infinite loop, this int array based check will be much faster, because you will detect the error after a few hundred iterations, so you will save 9000 string manipulations. The upper bound on the number of IDs is not the maximum possible values that the ID can represent but the number of items in the subject hierarchy, which in most cases is much smaller than 9999.

Another big issue with the 9999 is that who guarantees that it will be enough? Sure, usually it will be enough, but do you want to really add to the user documentation that maximum 9999 nodes can be exported using this feature?

The only advantage of the 9999 based solution is that it is a little bit simpler, but considering that it is less robust, less elegant and the performance improvement is either negligible or it actually makes things slower, I think we should not use this magic number based design. I'm open to other suggestions, but we can definitely do better than the current design.